File: Leetcode/338 - Counting Bits/solution.cpp
vector<int> countBits(int n) {
    vector<int> setBitCount(n + 1, 0);
    int exp = 1;

    for (int i = 1; i <= n; i++) {
        if (i == exp) {
            setBitCount[i] = 1;
            exp *= 2;
        } else {
            setBitCount[i] = setBitCount[exp >> 1] + setBitCount[i - (exp >> 1)];
        }
    }

    return setBitCount;
}

File: Leetcode/338 - Counting Bits/solution.ts
function countBits(n: number): number[] {
    const setBitCount: number[] = new Array(n + 1).fill(0);
    let exp: number = 1;

    for (let i = 1; i <= n; i++) {
        if (i === exp) {
            setBitCount[i] = 1;
            exp *= 2;
        } else {
            setBitCount[i] = setBitCount[exp >> 1] + setBitCount[i - (exp >> 1)];
        }
    }

    return setBitCount;
};